language/vm/src/deserializer.rs

```diff
@@ -1097,16 +1094,44 @@ fn load_function_def(cursor: &mut VersionedCursor) -> BinaryLoaderResult<Functio
     let flags = cursor.read_u8().map_err(|_| {
         PartialVMError::new(StatusCode::MALFORMED).with_message("Unexpected EOF".to_string())
     })?;
-    let is_public = (flags & FunctionDefinition::PUBLIC) != 0;
+
+    let (visibility, extra_flags) = match cursor.version() {
+        VERSION_1 => {
+            let vis = if (flags & Visibility::Public as u8) != 0 {
+                Visibility::Public
+            } else {
+                Visibility::Private
+            };
+            (vis, flags)
+        }
+        VERSION_2 => {
+            // NOTE: changes compared with VERSION_1
+            // - in VERSION_1: the flags is a byte compositing both the visibility info and whether
+            //                 the function is a native function
+            // - in VERSION_2: the flags only represent the visibility info and we need to advance
+            //                 the cursor to read up the next byte as flags
+            let vis = flags.try_into().map_err(|_| {
+                PartialVMError::new(StatusCode::MALFORMED)
+                    .with_message("Invalid visibility byte".to_string())
+            })?;
+            let extra_flags = cursor.read_u8().map_err(|_| {
+                PartialVMError::new(StatusCode::MALFORMED)
+                    .with_message("Unexpected EOF".to_string())
+            })?;
+            (vis, extra_flags)
+        }
+        _ => unreachable!("Invalid bytecode version"),
```

Comment:
return Err(PartialVMError::new(StatusCode::UNREACHABLE))

Best not to accidentally crash the VM :P